instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/ServerSpanNaming.java

```diff
@@ -58,15 +61,24 @@ public static void updateServerSpanName(
       }
       return;
     }
-    if (source.order > serverSpanNaming.updatedBySource.order) {
+    // special case for servlet filters, even when we have a name from previous filter see whether
+    // the new name is better and if so use it instead
+    boolean tryToFindBetterName =
+        !source.useFirst && source.order == serverSpanNaming.updatedBySource.order;
+    if (source.order > serverSpanNaming.updatedBySource.order || tryToFindBetterName) {
       String name = serverSpanName.get();
-      if (name != null) {
+      if (name != null && (!tryToFindBetterName || serverSpanNaming.isBetterName(name))) {
```

Comment:
I struggled to read this condition, I think this naming may help:
```suggestion
      if (name != null && (!onlyIfBetterName || serverSpanNaming.isBetterName(name))) {
```